File: 7_Merge two sorted lists/solution.cpp
#include <iostream>

struct ListNode{
    int val;
    ListNode* next;

    ListNode(): val(0), next(nullptr) {}
    ListNode(int x): val(x), next(nullptr) {}
    ListNode(int x , ListNode* next): val(x), next(next) {}

    void append(int val) {
        ListNode* new_node= new ListNode(val);
        ListNode* current = this;
        while(current->next != nullptr) {
            current = current->next;
        }
        current->next = new_node;
    }
    void print_list() {
        ListNode* current= this;
        while(current != nullptr) {
            std::cout << current->val << " ->" ;
            current = current->next;
        }
        std::cout << std::endl;
    }
};

ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
    ListNode* head = new ListNode(0);
    ListNode* current = head;

    while (list1 != nullptr && list2 != nullptr) {
        if(list1->val <= list2->val) {
            current->next = list1;
            list1= list1->next;
        } else {
            current->next = list2;
            list2 = list2->next;
        }
        current = current->next;
    }

    if (list1 != nullptr ) {
        current->next = list1;
    } else if (list2 != nullptr) {
        current->next = list2;
    }

    return head->next;
}

int main() {
    ListNode* list1 = new ListNode(1 , new ListNode(3 , new ListNode(5 )));
    list1->print_list();

    ListNode* list2 = new ListNode(1 , new ListNode(5 , new ListNode(7 )));
    list2->print_list();

    ListNode* result = mergeTwoLists(list1, list2); 

    result->print_list();

    return 0;
}

File: 7_Merge two sorted lists/solution.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def append(self, val):
        new_node = ListNode(val)
        current = self
        while current.next is not None :
            current = current.next
        current.next = new_node

    def print_list(self):
        while self:
            print(self.val ,end=' ->')
            self = self.next
        print("\n")
        
def mergeTwoLists(list1, list2):
    """
    :type list1: Optional[ListNode]
    :type list2: Optional[ListNode]
    :rtype: Optional[ListNode]
    """
    current1 = list1
    current2 = list2
    head = ListNode()
    current = head
    
    while current1 is not None and current2 is not None:
        if current1.val <= current2.val:
            current.next = current1
            current1 = current1.next

        else:
            current.next = current2
            current2 = current2.next

        current = current.next
    
    if current1 is not None:
        current.next = current1
    elif current2 is not None:
        current.next = current2
 

    return head.next



list1 = ListNode(1)
list1.append(2)
list1.append(4)
list1.print_list()

list2 = ListNode(4)
list2.append(8)
list2.append(7)

list2.print_list()
list3 = mergeTwoLists(list1, list2)
list3.print_list()